### PR TITLE
Training once per day

### PR DESCRIPTION
**Pull request by ibrahimsharaf**
- Issue (https://github.com/marco-c/crashsimilarity/issues/5)

### CODE REVIEW

File: utils.py [hunk 1]
```diff
@@ -14,17 +14,6 @@ def utc_today():
     return datetime.utcnow().date()
 
 
-def retrain_timeout_expired(training_interval=timedelta(hours=24)):
```
Comment (KirillTim) [inline utils.py]: Why do you delete this code?

File: crash_similarity.py [hunk 2]
```diff
@@ -76,10 +77,28 @@ def get_stack_trace_for_uuid(uuid):
     return data['proto_signature']
 
 
-def train_model(corpus):
-    if os.path.exists('stack_traces_model.pickle') and \
-       os.path.exists('stack_traces_model.pickle.docvecs.doctag_syn0.npy'):
-        return gensim.models.Doc2Vec.load('stack_traces_model.pickle')
+def delete_old_models(current_date, force_train):
```
Comment (KirillTim) [inline crash_similarity.py]: It is better to have "retrain timeout" as a variable parameter in any case. That's why `retrain_timeout_expired` function exists.

File: crash_similarity.py [hunk 3]
```diff
@@ -76,10 +77,28 @@ def get_stack_trace_for_uuid(uuid):
...
+    """
+    Get list of trained models inside the models directory,
+    delete all models in case of user forces new training,
+    if not, delete all models except of today's model
+    """
+    models = [f for f in os.listdir('./models')]
+    if force_train:
+        for model in models:
```
Comment (KirillTim) [inline crash_similarity.py]: Code duplication. You can use filter, or list comprehension here

**Comment by ibrahimsharaf** [comment]
@KirillTim I think using the current date (e.g: 14Apr2017) as part of the model's name, is simpler than writing training time in a file, retrieving it and calculating the interval, so I won't need this function in my case.

File: crash_similarity.py (hunk 3 above)
Comment (ibrahimsharaf) [inline crash_similarity.py]: Fixed.

File: crash_similarity.py [hunk 4]
```diff
@@ -83,14 +83,10 @@ def delete_old_models(current_date, force_train):
...
-    models = [f for f in os.listdir('./models')]
     if force_train:
-        for model in models:
-            os.remove('./models/' + model)
+        [os.remove('./models/' + model) for model in os.listdir('./models')]
     else:
-        for model in models:
-            if current_date not in model:
-                os.remove('./models/' + model)
+        [os.remove('./models/' + model) for model in os.listdir('./models') if current_date not in model]
```
Comment (KirillTim) [inline crash_similarity.py]: And what if we will want to change deletion logic? 
The idea is to have one line with `os.remove` and different logic of list creation. 
```models = [f for f in os.listdir('./models')]
if not force_train:
    models = [m for m in models if current_date not in m]
for m in models:
    os.remove('./models/{}'.format(m))
```

**Comment by KirillTim** [comment]
Anyway, why do you compare date as strings, but not time intervals?

**Comment by ibrahimsharaf** [comment]
`Anyway, why do you compare dates as strings but not time intervals?`

Simpler to implement.

**Comment by ibrahimsharaf** [comment]
@marco-c please review.

File: crash_similarity.py (hunk 4 above)
Comment (KirillTim) [inline crash_similarity.py]: @ibrahimsharaf There are still two places where `os.remove` is called. Fix it, please.

**Comment by KirillTim** [comment]
LPT: you can [squash](http://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git) commits to keep git log clean.

File: crash_similarity.py [hunk 5]
```diff
@@ -76,10 +77,25 @@ def get_stack_trace_for_uuid(uuid):
...
+    """
+    Get list of trained models inside the models directory,
+    delete all models in case of user forces new training,
+    if not, delete all models except of today's model
+    """
+    if force_train:
+        old_models = [model for model in os.listdir('./models')]
+    else:
+        old_models = [model for model in os.listdir('./models') if current_date not in model]
+    [os.remove('./models/{}'.format(model)) for model in old_models]
```
Comment (marco-c) [inline crash_similarity.py]: Can you make this a normal for loop?
```
for model in old_models:
    os.remove('./models/{}'.format(model))
```

File: crash_similarity.py [hunk 6]
```diff
@@ -231,9 +247,6 @@ def signature_similarity(model, paths, signature1, signature2):
     # download_data.download_crashes(days=7, product='Firefox')
     # paths = download_data.get_paths(days=7, product='Firefox')
     paths = ['crashsimilarity_data/firefox-crashes-2016-11-09.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-08.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-07.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-06.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-05.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-04.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-03.json.gz']
-
     corpus = read_corpus(paths)
-
     model = train_model(corpus)
-
-    print(dict([(model.index2word[i], similarity) for i, similarity in enumerate(model.similar_by_word('igdumd32.dll@0x', topn=False))])['igdumd64.dll@0x'])
+    print(dict([(model.wv.index2word[i], similarity) for i, similarity in enumerate(model.wv.similar_by_word('igdumd32.dll@0x', topn=False))])['igdumd64.dll@0x'])
```
Comment (marco-c) [inline crash_similarity.py]: Can you remove this and the previous unrelated changes? You can move this to a separate PR

File: crash_similarity.py (hunk 5 above)
Comment (ibrahimsharaf) [inline crash_similarity.py]: fixed.

File: crash_similarity.py (hunk 6 above)
Comment (ibrahimsharaf) [inline crash_similarity.py]: Done.

File: crash_similarity.py [hunk 7]
```diff
@@ -231,9 +248,6 @@ def signature_similarity(model, paths, signature1, signature2):
     # download_data.download_crashes(days=7, product='Firefox')
     # paths = download_data.get_paths(days=7, product='Firefox')
     paths = ['crashsimilarity_data/firefox-crashes-2016-11-09.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-08.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-07.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-06.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-05.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-04.json.gz', 'crashsimilarity_data/firefox-crashes-2016-11-03.json.gz']
-
```
Comment (marco-c) [inline crash_similarity.py]: Can you also remove these unrelated changes here (and the following two)?

**Comment by marco-c** [comment]
The tests are failing because the `models` directory doesn't exist.
You can add something like this at the beginning of the train function:
```
try:
    os.mkdir('models')
except OSError as e:
    if e.errno != errno.EEXIST:
      raise e
```

**Comment by ibrahimsharaf** [comment]
`Or add this code to utils.py in a makedir function and call it in the train function.`

Will we create a model directory if not found?, if not, is "makedir" the best name for the function?

**Comment by marco-c** [comment]
> Will we create a model directory if not found?, if not, is "makedir" the best name for the function?

It will create a directory if it doesn't exist, and will not fail if the directory already exists (while `os.mkdir` fails if the directory already exists).
I don't have any other name idea, but feel free to suggest a better name :)

File: crashsimilarity/crash_similarity.py [hunk 8]
```diff
@@ -70,10 +71,27 @@ def get_stack_trace_for_uuid(uuid):
...
-    if os.path.exists('stack_traces_model.pickle') and \
-       os.path.exists('stack_traces_model.pickle.docvecs.doctag_syn0.npy'):
-        return gensim.models.Doc2Vec.load('stack_traces_model.pickle')
+def delete_old_models(current_date, force_train):
+    """
+    Get list of trained models inside the models directory,
+    delete all models in case of user forces new training,
+    if not, delete all models except of today's model
+    """
+    utils.create_dir('models')
```
Comment (marco-c) [inline crashsimilarity/crash_similarity.py]: We should return early here if the directory doesn't exist, as if it doesn't exist, there will be no model to remove :)

File: crashsimilarity/crash_similarity.py [hunk 9]
```diff
@@ -96,7 +114,7 @@ def train_model(corpus):
     model.train(corpus)
     logging.info('Model trained in ' + str(time.time() - t) + ' s.')
 
-    model.save('stack_traces_model.pickle')
+    model.save('./models/stack_traces_' + current_date + '_model.pickle')
```
Comment (marco-c) [inline crashsimilarity/crash_similarity.py]: And you can create the directory here, before the call to save

Comment (ibrahimsharaf) [inline crashsimilarity/crash_similarity.py]: do we really need it? as the save line creates the directory itself.

Comment (marco-c) [inline crashsimilarity/crash_similarity.py]: If it creates the directory itself, then no, we don't need it!

File: crashsimilarity/crash_similarity.py (hunk 8 above)
Comment (ibrahimsharaf) [inline crashsimilarity/crash_similarity.py]: I think we can use something like [os.path.isdir](https://docs.python.org/dev/library/os.path.html#os.path.isdir), no need to add a new function in utils, right?

Comment (marco-c) [inline crashsimilarity/crash_similarity.py]: Yes, os.path.isdir works :)

File: crashsimilarity/crash_similarity.py (hunk 9 above)
Comment (ibrahimsharaf) [inline crashsimilarity/crash_similarity.py]: turns out it doesn't (my bad :sweat_smile:).

**Comment by marco-c** [comment]
Thanks!
